src/app/redux/auth/authActions.js

```diff
@@ -30,36 +35,41 @@ const setErrorWrongPassword = () => {
   }
 }
 
-const signIn = (credentials) => {
-  const uri = process.env.REACT_APP_BASE_API_URL + '/users?q=' + credentials.email;
+const setError = (msg) => {
+  return {
+    type: actionTypes.AUTH_ERROR,
+    error: msg,
+  }
+}
 
+const signIn = (credentials) => {
   return (dispatch) => {
-    fetch(uri)
-    .then(res => res.json())
-    .then(data => {
-      if(data.length === 0) {
-        dispatch(setErrorUserNotExist());
-      } else {
-        if(credentials.password !== data[0].password) {
-        dispatch(setErrorWrongPassword());
-        } else {
-          dispatch(setActiveUser({...data[0]}));
+    dispatch(setError(null));
 
-          localStorage.setItem("userEmail", data[0].email);
-          localStorage.setItem("userRole", data[0].role);
-          localStorage.setItem("userId", data[0].id);
-          localStorage.setItem("userName", data[0].name);
-        }
-      }
-    })
+    signInWithEmailAndPassword(authConnection, credentials.email, credentials.password)
+      .then(cred => {
+        getUserByEmail(cred.user.email)
+          .then(userData => {
+            if(userData !== null){
+              saveUserToBrowser({...userData});
+              dispatch(setActiveUser({ ...userData }));
+            }
+            else{
+              dispatch(setErrorUserNotExist());
+            }
+          })
+      })
```

Comment:
Is there any solution to avoid the loop of `then`?

Reply:
we can turn the returned function into an async function and use await instead of .then